src/dstack/_internal/utils/ignore.py

```diff
@@ -16,77 +17,83 @@ def __init__(
-            ignore_file = path / ignore_file
-            if ignore_file.exists():
-                self.load_ignore_file(str(path.relative_to(self.root_dir)), ignore_file)
+        # Collect patterns from all ignore files recursively
+        self._collect_patterns_recursive(self.root_dir, all_patterns)
 
-        for subdir in path.iterdir():
-            if not subdir.is_dir() or self.ignore(subdir.relative_to(self.root_dir)):
-                continue
-            self.load_recursive(subdir)
+        # Add additional glob patterns
+        all_patterns.extend(additional_globs)
 
-    @staticmethod
-    def rstrip(value: str) -> str:
-        end = len(value) - 1
-        while end >= 0:
-            if not value[end].isspace():
-                break
-            if end > 0 and value[end - 1] == "\\":
-                break  # escaped space
-            end -= 1
-        else:
-            return ""
-        return value[: end + 1]
+        self.parser = parse_gitignore_str("\n".join(all_patterns), self.root_dir)
 
-    @staticmethod
-    def fnmatch(name: str, pattern: str, sep="/") -> bool:
-        if pattern.startswith(sep):
-            name = sep + name
-        for n, p in zip_longest(
-            reversed(name.split(sep)), reversed(pattern.split(sep)), fillvalue=None
-        ):
-            if p == "**":
-                raise NotImplementedError()
-            if p is None:
-                return True
-            if n is None or not fnmatch.fnmatch(n, p):
-                return False
-        return True
+    def _collect_patterns_recursive(self, path: Path, patterns: List[str]):
+        """
+        Recursively collect patterns from all ignore files and combine them into a single gitignore,
+        with the root directory as the base path.
+        """
+        for ignore_file_name in self.ignore_files:
+            ignore_file = path / ignore_file_name
+            if ignore_file.exists():
+                try:
+                    # Get relative path from root to this directory
+                    if path == self.root_dir:
+                        prefix = ""
+                    else:
+                        prefix = path.relative_to(self.root_dir)
 
-    def ignore(self, path: PathLike, sep="/") -> bool:
-        if not path:
+                    # Read patterns and prefix them with directory path
+                    with ignore_file.open("r", encoding="utf-8", errors="ignore") as f:
+                        for line in f:
+                            line = line.strip()
+                            if line and not line.startswith("#"):
+                                if prefix:
+                                    # Prefix patterns with directory path for subdirectories
+                                    if line.startswith("/"):
+                                        # Absolute pattern within subdirectory
+                                        patterns.append(os.path.join(prefix, line[1:]))
```

Comment:
Slashes in the middle also indicate an absolute pattern.
> If there is a separator at the beginning or middle (or both) of the pattern, then the pattern is relative to the directory level of the particular `.gitignore` file itself.
[docs](https://git-scm.com/docs/gitignore#_pattern_format)

The previous implementation had the same flaw, but I think we can fix it now while we are at it.

```suggestion
                                    if "/" in line.rstrip("/"):
                                        # Absolute pattern within subdirectory
                                        patterns.append(os.path.join(prefix, line.lstrip("/")))
```